zyte_spider_templates/spiders/serp.py

```diff
@@ -117,12 +130,14 @@ def start_requests(self) -> Iterable[Request]:
         url = f"https://www.{self.args.domain.value}/search"
         for search_query in search_queries:
             search_url = add_or_replace_parameter(url, "q", search_query)
-            for start in range(0, self.args.max_pages * 10, 10):
-                if start:
-                    search_url = add_or_replace_parameter(
-                        search_url, "start", str(start)
-                    )
-                yield self.get_start_request(search_url)
+            yield self.get_serp_request(search_url)
 
     def parse_serp(self, response) -> Iterable[Serp]:
-        yield Serp.from_dict(response.raw_api_response["serp"])
+        serp = Serp.from_dict(response.raw_api_response["serp"])
+
+        next_start = serp.pageNumber * self._results_per_page
```

Comment:
I wonder if it'd be more reliable to pass the pageNumber around in the spider, instead of relying on extraction (which has a chance to fail)

Reply:
Somewhat related, I myself wonder if relying on the totalOrganicResults number is reliable as well, to be honest. I went that way to prioritize avoiding extra requests over reliability, but maybe that’s not a good idea.

I also considered counting organicResults, and only following when they are 10, but I have been told sometimes they could be lower than 10 in a regular page with follow up pages. Ignoring those issues, this approach would result in 1 extra request in cases where the last page has 10 results.

I also considered going to the next page if organicResults are 0. This has the advantage that pages with 1 or 2 fewer organic results will not cause pagination to stop, and the disadvantage that you reliably get 1 extra empty results page every time.

Any thoughts on this?

Reply:
I think we can keep the current approach for totalOrganicResults for now, and change it if there are issues with it. It seems there are tradeoffs, and it's not clear what's the best approach for stopping in practice. The case for 0 results is handled in the current code, which is good.